### ISSUE TITLE
Support Jira issues

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#3796]
**What would you like Renovate to be able to do?**

Use Jira for Bitbucket Server for issues (e.g. config error notification, Master Issue, etc).

**Describe the solution you'd like**

Ideally Jira can be autodiscovered from Bitbucket Server and same credentials used, but otherwise we need additional configuration.

It is not yet known if this is feasible.

**Comment by rarkins** [comment]
Do any Bitbucket Server users like @iamstarkov or @ViceIce have prior knowledge about this?

**Comment by ViceIce** [comment]
I've looked at it before, cause i've had the same idea. This should generally work, but jira don't use markdown, they have their own markup language

**Comment by ViceIce** [comment]
https://community.atlassian.com/t5/Jira-questions/Using-markdown-in-Jira/qaq-p/892041

**Comment by ViceIce** [comment]
[Markin - Markdown for Jira Cloud/Server](https://marketplace.atlassian.com/apps/1214558/markin-markdown-for-jira-cloud-server?hosting=server&tab=overview) could be an option to support markdown

**Comment by rarkins** [comment]
I'm not too worried about converting our markdown to Jira format if we need to though. It should be a reasonably easy one-way translation

**Comment by ViceIce** [comment]
maybe this should be configurable for users, becuase the field render can be changed on project base.

so we have to configure the corresponding project anyway, because there is no direct link between a jira project and a bitbucket repository

**Comment by rarkins** [comment]
So we can't "discover" an attached Jira server automatically - we'd need that configured in addition to the VCS endpoint? And perhaps also an additional username/password (token) if they are not shared across systems?

**Comment by ViceIce** [comment]
not as i'm aware of. [rest apis](https://developer.atlassian.com/server/bitbucket/reference/rest-api/)

maybe there is some hidden rest api

**Comment by psyb0t** [comment]
@ViceIce did you ever get the issue tracker enabled on the repos on bitbucket server? i'm running bitbucket server and jira in docker and connected bitbucket with jira during the setup process and i don't seem to find anything related to issues.

**Comment by ViceIce** [comment]
You will see it, if any issue id is in commit message or pr 

on pr:
![image](https://user-images.githubusercontent.com/1798109/58165463-23d3f280-7c88-11e9-8959-b6e7cfd2c4cf.png)

on commit list:
![image](https://user-images.githubusercontent.com/1798109/58165548-482fcf00-7c88-11e9-9f24-10e96bc636c5.png)

i've added `commitBody: '[VO-113]'` to my `config.js`

**Comment by ViceIce** [comment]
on an jira issue:
![image](https://user-images.githubusercontent.com/1798109/58166209-9abdbb00-7c89-11e9-97d2-c72d17a181b0.png)

**Comment by ViceIce** [comment]
VSAF-113 is my currently used virtual renovate master issue 😄

**Comment by ViceIce** [comment]
> So we can't "discover" an attached Jira server automatically - we'd need that configured in addition to the VCS endpoint? And perhaps also an additional username/password (token) if they are not shared across systems?

Found a possibillity (maybe undocumented)

`GET /rest/jira-integration/latest/servers`

Response:
```json
[
    {
        "id": "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX",
        "name": "JIRA Sample",
        "displayUrl": "https:\/\/jira.example.com",
        "rpcUrl": "https:\/\/jira.example.com",
        "authenticated": true,
        "selected": true
    }
]
```

We can use this to find jira and enable issues.

**Comment by ViceIce** [comment]
Jira supported text formatings: 
[rhorber_jira-text-formatting-notation.pdf](https://github.com/renovatebot/renovate/files/3387565/rhorber_jira-text-formatting-notation.pdf)

**Comment by ps-jay** [comment]
Thinking about this, since the `masterIssue` looks like a dashboard ([reference blog post](https://renovate.whitesourcesoftware.com/blog/introducing-renovates-master-issue/)), I wonder if Confluence might be a better spot this kind of thing?

But then thinking more broadly, maybe this could be generalised for any kind of wiki?

Complementary to `masterIssue`, perhaps there could be an option for `wikiDashboard` ?
This would then have its' own set of properties, e.g.
* `wikiPlatform`
* `wikiEndpoint`
* `wikiParent` (e.g. store all the repo's dashboard pages under this page)

Alternative thinking, what if we could output README markdown files into a user configurable repo?  Then let Bitbucket render the README.md files
e.g. `renovate-bot-dashboard-repo/FOOPROJECT/bar-repo/README.md`

Any of these ideas seem worth while?

**Comment by rarkins** [comment]
The `masterIssue` is most powerful when it's in the same repo and the checkboxes work. Without that, it's of much less value. The biggest motivator for this would be for alerting to things like config errors.

**Comment by ps-jay** [comment]
Of course, I hadn't considered the interactivity side of it.

In that case, README markdown files aren't the way to go at all.

Maybe it does make sense to try and scope this to Confluence then, since it has a checkbox feature.
It would seem a better fit for this kind of dashboard than Jira for these reasons:
- there's not a 1:1 match of Jira projects to Bitbucket repos in all the organisations I've worked at
- usually, it's not good to have a long running, never completed story in Jira

**Comment by houda95** [comment]
Would also love to see a similar feature for integrating Renovate with Bitbucket cloud. Having this functionality would definitely help Renovate adoption within my team.

**Comment by rarkins** [comment]
> Would also love to see a similar feature for integrating Renovate with Bitbucket cloud. Having this functionality would definitely help Renovate adoption within my team.

How would you suggest?

**Comment by Mcluky** [comment]
Hi, I'm new to the Renovate community but stumbled upon this Issue. We're currently in the process of setting up Renovate in our company for Bitbucket Server and searched for this feature to create a new Jira Issue when dependencies need an update. I hope I'm in the right place for this.
I've got some experience in developing with Atlassian software in the sense that I've been working with the Atlassian ecosystem (Jira/Confluence/Bitbucket ) for the last year. I've had a look at this and found a REST-API that allows us to create issues through Bitbucket on Jira ([https://docs.atlassian.com/bitbucket-server/rest/7.4.0/bitbucket-jira-rest.html#idp2l](https://docs.atlassian.com/bitbucket-server/rest/7.4.0/bitbucket-jira-rest.html#idp2l)). The API works with the existing credentials that are required for bitbucket. As @viceice has already pointed out, you can discover the required Jira instance also using a REST call. For me, the URL was slightly different (``GET rest/jira-integration/1.0/servers``, same response) and I found no official documentation for this. I see that there are still some tricky things to consider. For example, you are required to have the project key for Jira but in general, I'm very optimistic that we could implement this.
I'm willing to invest some time into this as soon as we've got Renovate running for us.

**Comment by rarkins** [comment]
@Mcluky we'd be very happy to have you contribute!

Probably the biggest question to answer first though is: what should be the scope of functionality? You may need to experiment with Renovate for a while before you form an opinion.

With GitHub and GitLab issues, we can use them to:
- Alert the user if a critical config error prevents the running of Renovate (without this, users may not be aware that dependencies are no longer being updated)
- Create issues for deprecated dependencies (this functionality is currently on hold as some bugs or misunderstandings in the initial code lead to duplicate issues being spammed in some scenarios)
- Creating the "Dependency Dashboard"

The Dashboard is super useful, but it makes use of a "trick" with markdown checkboxes to achieve interactive functionality. I'm not sure if any same trick can be used in Jira, so then it would be a read-only dashboard instead.

**Comment by nickstolwijk** [comment]
We are using Renovate on Bitbucket server and Jenkins for a while now. We see some problems with our renovate.json configurations but we are not really getting notified of them. No one reads the Jenkins buildlog which executes Renovate and there is no notification on the repositories having the error.

Example errors:
- A local preset which does not exist anymore (renamed)
- package rules which are not valid (could have been picked up by renovate-config-validator but not everyone uses that locally)

I was hoping the dependency dashboard should show us these issues, but I found that that functionality was not implemented for bitbucket-server.

A few things I missed in the documentation:
- I did not find any mention of the configuration errors in the documentation of the dependency dashboard https://docs.renovatebot.com/configuration-options/#dependencydashboard
- I am missing what functionality should or should not work for bitbucket-server
https://docs.renovatebot.com/modules/platform/

What I would like to see:
- the configuration errors for a repository close to that repository (instead of the renovate logging).

Possible solution (for bitbucket-server):
- Create a PR with the logging in the repository which caused the warning, assigning the reviewers from renovate.json to that PR, so they can see that something goes amiss in their repository.

**Comment by Mause** [comment]
I did some investigation on this in #16167, and looking at the not-really-documented API's, it should be possible. The API's themselves are "documented" in the WADL format at `/rest/jira-integration/1.0/application.wadl` on the instance itself.

The `/rest/jira-integration/1.0/servers` returns information about the linked servers, including jira, and gives the URL. It's possible that we could just POST to `/rest/jira-integration/1.0/issues` (as documented in the WADL), but that might require doing some OAuth linking between the service accounts on both systems

**Comment by setchy** [comment]
@rarkins could you kindly add the `platform:bitbucket` label to this issue 🙏

**Comment by rarkins** [comment]
I think Bitbucket cloud and Jira cloud are a popular combination. Ideally Jira issue support would eventually be platform independent though

**Comment by viceice** [comment]
> I think Bitbucket cloud and Jira cloud are a popular combination. Ideally Jira issue support would eventually be platform independent though 

agreed, so we should first discuss a solution and then open a separate issue with conclusion.

this issue is about support Jira server for bitbucket server, so we shouldn't mix those things.

**Comment by rarkins** [comment]
I've updated the title to remove some ambiguity. @setchy thanks for moving this long-awaited feature forward!

**Comment by setchy** [comment]
As promised, https://github.com/renovatebot/renovate/issues/20568 created for the Jira Cloud + Bitbucket Cloud combo

**Comment by Shuyinsama** [comment]
As an interesting side not. Atlassian is currently on it's way to fully drop support for the self-hosted (server) options for all their services: https://www.atlassian.com/migration/assess/journey-to-cloud

It is heavily recommended to migrate to either their cloud services or something else entirely.
Is this issue still worth the investment then?

I understand a lot of companies will probably keep using Jira server (and bitbucket server) in the coming years but you will miss out on any security and feature updates.

Is it worth developing support for an already ending service?

**Comment by braunsonm** [comment]
> Is it worth developing support for an already ending service?

This statement is not true. Atlassian data center is still supported and not going away. They just removed the server tier.

Issue should probably be renamed but it is effectively the same product and API

**Comment by ps-jay** [comment]
They are ending support for the (relatively) low cost server products

Self-hosted editions live on as the more expensive data-centre product line

**Comment by Shuyinsama** [comment]
Ah ok. I was under the impression that server and data center were "different" in terms of API's etc.
Then yes this would still be a good integration! Thanks for the correction!

**Comment by HenrikPoulsen** [comment]
I'm not sure why this issue is specifically bit bucket + jira? Would it not make sense to just make jira a configurable option to replace the normal issue creation?

My use case is that most github repos at my company are disabled entirely, since we use jira due to there being a lot of cross team/org collaboration as well as multiple github instances etc.

So for those repos Renovate issues don't get created anywhere, making it harder to see what the problem is.

**Comment by viceice** [comment]
@HenrikPoulsen we've a different issue for that feature request.

**Comment by rarkins** [comment]
I agree that Jira Issues (whether cloud or data center) should be configurable independently of the platform

**Comment by HenrikPoulsen** [comment]
I lust be blind because it can't really find the generic Jira issue

**Comment by viceice** [comment]
> I lust be blind because it can't really find the generic Jira issue

- https://github.com/renovatebot/renovate/issues/20568

see discussion there

**Comment by reitzig** [comment]
I'd like to go on record with saying that even a non-interactive dashboard issue that just shows the current state of Renovate-things (of the given repo, one per issue?) would be a _huge_ help.
